### PR TITLE
Add Benchmark for Vectorized Parquet Read

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?
#536 task 7

testing environment:
Java HotSpot(TM) 64-Bit Server VM 1.8.0_144-b01 on Linux 2.6.32_1-18-0-0
Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz

Add Some Benchmark code to compare:

1. FullScan 

- ParquetReader MR (base) 

- OapParquetReader Vectorized  -> Batch 

- ParquetReader Vectorized     -> Batch

- OapParquetReader Vectorized  -> Row

- ParquetReader Vectorized     -> Row 

2.IndexedScan -> 

- degraded Index MR reader(base) 

- degraded Indexed  Vectorized reader 

- Indexed MR reader

- Indexed  Vectorized reader 

## How was this patch tested?

not require testing

This work not complete, I'll add more case.

### CODE REVIEW

**Comment by LuciferYang** [comment]
Parquet Reader Single Int Column Scan:   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
        Oap ParquetReader MR                          1544 / 1583         10.2          98.2       1.0X
        OapParquetReader Vectorized  -> Batch         130 /  155         121.4           8.2      11.9X
        ParquetReader Vectorized     -> Batch         142 /  169         111.2           9.0      10.9X
        OapParquetReader Vectorized  -> Row           217 /  255          72.6          13.8       7.1X
        ParquetReader Vectorized     -> Row           235 /  265          66.9          14.9       6.6X

**Comment by LuciferYang** [comment]
fractionOfNulls = 0.95
        String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
        ------------------------------------------------------------------------------------------------
        Oap Parquet MR                                1188 / 1193          8.8         113.3       1.0X
        Oap Parquet Vectorized                         172 /  189         61.1          16.4       6.9X
        Parquet Vectorized                             184 /  205         56.9          17.6       6.4X
        Oap Parquet Vectorized (Null Filtering)        171 /  173         61.4          16.3       7.0X
        Parquet Vectorized (Null Filtering)            183 /  187         57.2          17.5       6.5X

**Comment by LuciferYang** [comment]
fractionOfNulls = 0.50
        String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
   ------------------------------------------------------------------------------------------------
        Oap Parquet MR                                1922 / 1955          5.5         183.3       1.0X
        Oap Parquet Vectorized                         876 /  887         12.0          83.6       2.2X
        Parquet Vectorized                             897 /  938         11.7          85.5       2.1X
        Oap Parquet Vectorized (Null Filtering)        902 /  932         11.6          86.1       2.1X
        Parquet Vectorized (Null Filtering)            912 /  940         11.5          86.9       2.1X

**Comment by LuciferYang** [comment]
fractionOfNulls = 0.00
        String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
   ------------------------------------------------------------------------------------------------
        Oap Parquet MR                                2346 / 2368          4.5         223.8       1.0X
        Oap Parquet Vectorized                        1120 / 1130          9.4         106.8       2.1X
        Parquet Vectorized                            1134 / 1138          9.3         108.1       2.1X
        Oap Parquet Vectorized (Null Filtering)       1005 / 1023         10.4          95.8       2.3X
        Parquet Vectorized (Null Filtering)           1044 / 1062         10.0          99.5       2.2X

**Comment by LuciferYang** [comment]
Indexed Sql Single Int Column Scan:      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
   ------------------------------------------------------------------------------------------------
        SQL Parquet MR -> skipIndex                   2553 / 2577          6.2         162.3       1.0X
        SQL Parquet Vectorized -> skipIndex            189 /  198         83.2          12.0      13.5X
        SQL Parquet MR -> force useIndex               578 /  581         27.2          36.7       4.4X
        SQL Parquet Vectorized -> force useIndex       158 /  168         99.5          10.0      16.2X

**Comment by LuciferYang** [comment]
Copy result to commet and It show in code commets too.

**Comment by linhong-intel** [comment]
@yueawang to compare with our own benchmark result

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/IndexedParquetReadBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        benchmark.run()
+      }
+    }
+  }
+
+  def main(args: Array[String]): Unit = {
+    intScanBenchmark(1024 * 1024 * 15)
+  }
+}
+
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/IndexedParquetReadBenchmark.scala]: Nit: remove 1 line from end

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+  }
+
+  def main(args: Array[String]): Unit = {
+    for (fractionOfNulls <- List(0.95, 0.50, 0.0)) {
+      stringWithNullsScanBenchmark(1024 * 1024 * 10, fractionOfNulls)
+    }
+    intScanBenchmark(1024 * 1024 * 15)
+  }
+}
+
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala]: Nit: remove 1 line from end

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/BenchmarkUtils.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.util.Utils
+
+object BenchmarkUtils {
+
+  private val conf = new SparkConf()
+  conf.set("spark.sql.parquet.compression.codec", "snappy")
+  conf.set("spark.memory.offHeap.size", "200m")
+
+
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/BenchmarkUtils.scala]: Nit: remove 1 empty line

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/BenchmarkUtils.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  val spark: SparkSession = SparkSession.builder
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/BenchmarkUtils.scala]: Nit: break after `=` is prior to break around `.`

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/BenchmarkUtils.scala [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    .master("local[1]")
+    .appName("test-oap-context")
+    .config(conf)
+    .getOrCreate()
+
+  // Set default configs. Individual cases will change them if necessary.
+  spark.conf.set(SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key, "true")
+  spark.conf.set(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key, "true")
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/BenchmarkUtils.scala]: Are these just default values if not set?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+      withTempTable("t1", "tempTable") {
+        spark.range(values).createOrReplaceTempView("t1")
+        spark.sql(s"select IF(rand(1) < $fractionOfNulls, NULL, cast(id as STRING)) as c1, " +
+          s"IF(rand(2) < $fractionOfNulls, NULL, cast(id as STRING)) as c2 from t1")
+          .write.parquet(dir.getCanonicalPath)
+        spark.read.parquet(dir.getCanonicalPath).createOrReplaceTempView("tempTable")
+
+        val benchmark = new Benchmark("String with Nulls Scan", values)
+
+
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala]: Nit: too much empty line

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+                val value = row.getUTF8String(0)
+                if (!row.isNullAt(0) && !row.isNullAt(1)) sum += value.numBytes()
+              }
+            } finally {
+              reader.close()
+            }
+          }
+        }
+
+
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala]: Nit: too much empty line

**Comment by yueawang** [comment]
LGTM. Generally our oap benchmark suite gives out same result, i.e.  oap parquet vectorized ~= parquet vectorized >> oap parquet base.

**Comment by LuciferYang** [comment]
@adrian-wang Thank u for ur reivew~ already format code which u pointed out~

**Comment by LuciferYang** [comment]
@yueawang For `parquet W/O Index` , I think the result of  03_14 quick than 03_05 because of #583.
`parquet w/ index` slow than `parquet W/O Index`  perhaps because of the extra overhead of the IndexScanners.indexIsAvailable method, It will bring about 300ms overhead for each task(test by our qa).

**Comment by LuciferYang** [comment]
@yueawang On the other hand, can u tell me how many indexScanner have been skiped of `parquet w/ index` case?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+                val rowIterator = batch.rowIterator()
+                while (rowIterator.hasNext) {
+                  sum += rowIterator.next().getUTF8String(0).numBytes()
+                }
+              }
+            } finally {
+              reader.close()
+            }
+          }
+        }
```
Comment (gczsjdy) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala]: Can we split this to multiple functions and maybe extract some common parts?

Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/benchmark/FullScanParquetReaderBenchmark.scala]: Now, this work has not been completed, after complete all code commit, I can try to do this refactor work :)

**Comment by linhong-intel** [comment]
close it first.

**Pull request closed without merging**